Make handling of non-ASCII keys more consistent

vadz opened the pull request:
This is a backwards incompatible change, but I think we can make it in 3.3.0 and it should make the behaviour as consistent and logical as possible: the key codes are now always ASCII characters (and not things like `ù` or `²` for example) that correspond either to the actual value of the key or to the value of the same physical key in US layout if it doesn't produce an ASCII character in the current layout.

Any comments from people, especially those using non-US Latin keyboard layouts (this PR doesn't change much for non-Latin layouts like Cyrillic ones) would be very welcome!

vadz committed:
Generate key codes in US layout for OEM keys in wxMSW

Instead of not generating any key code at all for such keys, generate
the key code that the same key would produce in the US layout.

This is especially useful for the keyboard using Cyrillic layout, as
they would typically have both sets of characters on them and it makes
sense to generate an event with the key code of e.g. `;` for the key
with this marking even if it produces a Cyrillic letter when pressed.

Update the documentation to document this behaviour, which was already
implemented in wxGTK, and generally try to explain things more clearly.

Note that this doesn't affect accelerators, which are handled by
Windows, i.e. this isn't sufficient to fix #22058 under MSW.

vadz committed:
Allow triggering more accelerators in non-US keyboard layouts

If the character used in the accelerator (e.g. "]" used in "Ctrl-]") is
not present in the layout, use the key that would have generated it in
the US layout instead.

This is arguably better than nothing, but it still doesn't allow to
generate all the possible key combinations, so update the documentation
to warn about it.

vadz committed:
Generate key codes consistent with wxMSW for VK_OEM_xxx in wxGTK

These keys should generate the key code corresponding to their label on
the physical keyboard if it's a Latin-1 character both because this is
how wxMSW behaves and because this is likely what the user expects.

Do this by using the Unicode character corresponding to one of these
keys if it's a Latin-1 one, just as we already did it for the letters.

vadz commented:
@unxed Will you be able to test this?

taler21 commented:
> Any comments from people, especially those using non-US Latin keyboard layouts

I tested it with a German keyboard layout and can confirm that it works as described.

However, I cannot see the benefit (for non-US Latin keyboard layouts).

Until now, with a German keyboard layout, the key code of the keys `Ä`, `Ö`, `Ü` corresponded to the physical keys (`ä`, `ö`, `ü`), i.e. the labeling of the keys visible to the user.
But with this PR, the key code of the keys `Ä`, `Ö`, `Ü` would be `'`, `;`, `[`.

In the previous scenario, as it works so far at least under Windows and Linux, an accelerator like `Strg+ü` would be possible (although not recommended) in a German application.
By the way, Visual Studio in the German edition actually uses such an accelerator (`Umschalt+Alt+ü`).

But if the key code represents the physical key of the US keyboard instead, it becomes useless.
Because an accelerator with a key that is labeled differently on the keyboard is not understandable for users.

However, I am not decidedly against this change, since we do not use exotic (non-ASCII) accelerators in our applications anyway.
But the previous behavior seems to make more sense to me (for non-US Latin keyboard layouts).

vadz commented:
> > Any comments from people, especially those using non-US Latin keyboard layouts
> 
> I tested it with a German keyboard layout and can confirm that it works as described.

Thanks for testing!
 
> However, I cannot see the benefit (for non-US Latin keyboard layouts).
> 
> Until now, with a German keyboard layout, the key code of the keys `Ä`, `Ö`, `Ü` corresponded to the physical keys (`ä`, `ö`, `ü`), i.e. the labeling of the keys visible to the user. But with this PR, the key code of the keys `Ä`, `Ö`, `Ü` would be `'`, `;`, `[`.

I had originally made this change for the French layout in which we were getting key events for `²` which seemed just weird. But I admit that this logic works much less well for the German layout, so I'll probably revert 34345e2343 (Change key code for non-ASCII Latin-1 characters in wxMSW, 2025-03-30)...

vadz commented:
P.S. One thing which _really_ doesn't make sense even for German is that the key code is `ä` and not `Ä`, which is inconsistent with all the other letters. OTOH we couldn't generate the uppercase version of `ß` which doesn't exist and I don't know if it's worth breaking compatibility for `ù`, `ä` etc...

unxed commented:
Will take look at it tomorrow

vadz commented:
> Will take look at it tomorrow

Thanks, but please note that I'll push a commit reverting the handling of high-ASCII characters soon (not sure if today, however).

vadz committed:
Further improve documentation of KEY event key codes

Document that non-ASCII letters don't get converted to upper case.

Explain that the strange behaviour is due to compatibility reasons and
give an example separately instead of doing it in the middle of the text
explaining the rules.

vadz committed:
Change the code for finding non-Latin key codes in wxMSW

Make it more general, in case we decide that we want to generate these
key codes for non-ASCII (and not just non-Latin-1) characters in the
future, and arguably more clear, even if longer, as we now hardcode the
US keyboard layout.

No real changes in behaviour.

vadz commented:
> > Will take look at it tomorrow
> 
> Thanks, but please note that I'll push a commit reverting the handling of high-ASCII characters soon (not sure if today, however).

I did do it today, finally. This should be ready for testing now.

vadz commented:
Sorry, I've just realized that I didn't update wxGTK behaviour to match the new (old) wxMSW behaviour, so I've just done it now. Hopefully you haven't tested under Linux yet, but if you did, please retest with the latest version because in it wxGTK also generates `ù` and `ä` etc key codes, as wxMSW does.

unxed commented:
today was too long day unfortunately. will try to test it tommorow or a day after

unxed commented:
And what about Macs? Unfortunately have no Mac to test

taler21 commented:
I tested it again with the German keyboard layout and can confirm that it now works again as with 3.2.

> One thing which really doesn't make sense even for German is that the key code is `ä` and not `Ä`, which is inconsistent with all the other letters.

Yes, that's what I thought too, since the physical keys `Ä`, `Ö`, `Ü` on the keyboard are also labeled in capital letters.

> OTOH we couldn't generate the uppercase version of ß which doesn't exist and I don't know if it's worth breaking compatibility for `ù`, `ä` etc...

Probably not worth breaking compatibility. Especially since Visual Studio also uses a lowercase `ü` in its accelerator (`Umschalt+Alt+ü`).

vadz commented:
As I haven't had any other negative feedback, I'm going to merge this soon.

unxed commented:
You'll laugh, but I only just now found a free minute. I'll check it out, then, already in the master - better than nothing
